### PR TITLE
Reports: Cohort quarter improvements

### PR DESCRIPTION
**Pull request by rsanheim**
**Story card:** https://app.clubhouse.io/simpledotorg/story/779/add-5th-quarter-cohort-data

* add the slanted bars for 1st quarter
* show 5th quarter
* show the dash in the Quarter string
* improve caching for the CohortService

image

### CODE REVIEW

**Comment by claudiovallejo** [comment]
Looks great Rob!

File: app/presenters/user_analytics_presenter.rb [hunk 1]
```diff
@@ -226,13 +226,8 @@ def all_time_stats
...
   def cohort_stats
-    cohort_cache_version = 1
-    cohort_cache_key = "user_analytics/#{current_facility.id}/cohort_stats/#{cohort_cache_version}"
-
-    Rails.cache.fetch(cohort_cache_key, expires_in: 7.days) do
-      quarters = Quarter.new(date: Date.current).previous_quarter.downto(3)
-      CohortService.new(region: current_facility, quarters: quarters).call
-    end
+    quarters = Quarter.new(date: Date.current).previous_quarter.downto(3)
+    CohortService.new(region: current_facility, quarters: quarters).call
```
Comment (rsanheim) [inline app/presenters/user_analytics_presenter.rb:230]: @harimohanraj89 The caching is done in the service now, so both callers of this service should benefit from it.

File: app/services/cohort_service.rb [hunk 2]
```diff
@@ -11,29 +13,45 @@ def initialize(region:, quarters: nil)
...
+      quarter_data = compute_quarter(results_quarter)
+      result[:quarterly_registrations] << quarter_data
+    end
+    result
+  end
 
+  private
+
+  def compute_quarter(results_quarter)
+    Rails.cache.fetch(cache_key(results_quarter), version: cache_version, expires_in: CACHE_TTL, force: force_cache?) do
```
Comment (rsanheim) [inline app/services/cohort_service.rb:25]: Caching each quarter result set.

File: spec/services/cohort_service_spec.rb [hunk 3]
```diff
@@ -23,12 +24,12 @@ def refresh_views
...
-      create(:patient, registration_facility: facility, recorded_at: jan_5 + 20.days),
-      create(:patient, registration_facility: facility, recorded_at: jan_5 + 30.days),
-      create(:patient, registration_facility: facility, recorded_at: jan_5 + 45.days),
-      create(:patient, registration_facility: facility, recorded_at: jan_5 + 60.days)
+      create(:patient, registration_facility: facility, registration_user: user, recorded_at: jan_5),
+      create(:patient, registration_facility: facility, registration_user: user, recorded_at: jan_5 + 10.days),
+      create(:patient, registration_facility: facility, registration_user: user, recorded_at: jan_5 + 20.days),
+      create(:patient, registration_facility: facility, registration_user: user, recorded_at: jan_5 + 30.days),
+      create(:patient, registration_facility: facility, registration_user: user, recorded_at: jan_5 + 45.days),
+      create(:patient, registration_facility: facility, registration_user: user, recorded_at: jan_5 + 60.days)
```
Comment (rsanheim) [inline spec/services/cohort_service_spec.rb:32]: I wish we had an easier way to have large amounts of fixture data use the same associated objects ...saves a lot on setup time.

File: app/presenters/user_analytics_presenter.rb (hunk 1 above)
Comment (harimohanraj89) [reply app/presenters/user_analytics_presenter.rb:230]: Nice. The `RegionReportService` was doing some funky organization-IDs-in-the-cache-key stuff, so I left it untouched. Thanks for consolidating! 👍  I'm guessing we're no longer going to support different users seeing different reports for the same region based on their organizational access?

File: spec/services/cohort_service_spec.rb (hunk 3 above)
Comment (harimohanraj89) [reply spec/services/cohort_service_spec.rb:32]: Hmm, yeah. I can't find a FactoryBot way to do this. We _could_ pull the common attributes into a hash and splat-include it in each line here.

```ruby
associations = { registration_facility: facility, registration_user: user }

create(:patient, recorded_at: jan_5, **associations),
# ...
```

Ideally, we want some sort of `create_with() do` thing 🤔

**Pull request merged**